### PR TITLE
login mechanism using trusted SSO headers as describ…

### PR DESCRIPTION
**Pull request by verybadsoldier**
…ed here:

https://www.authelia.com/integration/trusted-header-sso/introduction/

## What type of PR is this?
- feature

## What this PR does / why we need it:

Rough draft implementation of the concept of trusted header SSO as described here:
https://www.authelia.com/integration/trusted-header-sso/introduction/

This PR is only to demonstrate the basic implementation concept. Lots of polishing will be needed to get it finished. I have no experience with Go, Vue or Typescript. So handle with care.

The idea is:
* Add a new API endpoint `login-sso-header` to login via HTTP SSO headers
* `login-sso-header` will generate an access token in the same way the regular `/login` endpoint does
* Everytime before showing the login page the frontend will access `login-sso-header` to see if it can login using SSO
* If this succeeds then the frontend will redirect to `/home` (just like with regular login)
* In this demo the HTTP header `Remote-Email` will be used. An account belongig to that email address has to exist already. No automatic creation of accounts right now.

## Which issue(s) this PR fixes:
https://github.com/hay-kot/homebox/issues/270
https://github.com/hay-kot/homebox/issues/331

## Special notes for your reviewer:
This is a draft PR. So it is not actually meant to get merged. I would like to ask you for your opinion about this approach. As I have little experience with web applications I am not sure if this is the best approach to do it. Please advise if there are better way to do it.

### CODE REVIEW

**Comment by verybadsoldier** [comment]
Here is some progress in another branch that adds some configuration for it:
https://github.com/verybadsoldier/homebox/tree/header_sso_config

Usage:
```
HBOX_OPTIONS_HEADER_SSO_ENABLED=1
HBOX_OPTIONS_HEADER_SSO_ALLOWED_IP=127.0.0.1
```
Enables ability to login by SSO headers and the second one needs to be set to the IP of the reverse proxy for security reason, so no random clients can use the trusted headers.

**Comment by verybadsoldier** [comment]
Integrated the config parameters into this branch and removed draft status.

**Review by hay-kot — COMMENTED** [review]
This looks like a good start. I have a few questions.

1. How does the SSO user correspond to a user in the Homebox Database and/or how are groups handled? (or what's the plan there since it's not implemented)
2. How does this work in practice? What happens if I just navigate to the /login directly without first going through the auth portal? Does there need to be a button that's like "Sign in with SSO" to redirect them to the Authelia portal (or whatever)? - I guess I'm thinking that if SSO is enabled, we'd probably want to hide the login all together, right? 🤷

**Comment by verybadsoldier** [comment]
Well, it works much simpler than one might think. It is really just an _additional_ way to obtain an access token. User or group management stays untouched. Even when enabled, the normal functionality is not altered or degraded in any way.
Enabling SSO only has an effect if someone sends a request to the index page (/) and providing the additional HTTP header `Remote-Email` in the request.

> 1. How does the SSO user correspond to a user in the Homebox Database and/or how are groups handled? (or what's the plan there since it's not implemented)

The two user database (Homebox + SSO service) are totally independent of each other. A user has to be created manually in Homebox as in the past (and also in the SSO database). Also, group assignment for this user happens in Homebox. The only connection is that it is expected to have a user in the SSO service with the same email address as a user in Homebox (email address is used for the matching).

What can be done (and maybe should be done): there could be a mechanism that automatically creates users in Homebox. The set of SSO headers contains `Remote-User`, `Remote-Groups`, `Remote-Name` and `Remote-Email`. So when these headers are provided and no user matching `Remote-Email` can be found in Homebox then a user could automatically be created using the data from all headers. That way, no users would have to be created manually in Homebox when using SSO.
I can offer to provide an implementation for that (I played around with it and I think it is not too complicated).

> 2. How does this work in practice? What happens if I just navigate to the /login directly without first going through the auth portal? Does there need to be a button that's like "Sign in with SSO" to redirect them to the Authelia portal (or whatever)? - I guess I'm thinking that if SSO is enabled, we'd probably want to hide the login all together, right? 🤷

Hm, as I understand it, `/login` is an API endpoint, so a user would not navigate (with a browser) to it? You mean navigating to the regular start page (index)?
If SSO is enabled but the index page is requested without going through an SSO service, then no SSO headers (`Remote-Email` etc.) will be available so the SSO authentication will just fail. In that case, the regular login page will be presented, and the user can log in the traditional way (user email/password to the regular Homebox user database).

In my opinion, we don't have to hide the regular login page as both SSO login and regular login can live in perfect coexistence (as described above) and the regular login page can still act as a fallback.
In my understanding, such a button "Sign in with SSO" is only needed when using something like OpenID where you navigate first to the client login page (Homebox) and then such a button redirects you to the SSO login. But well, my experience is also limited with that.

**Comment by verybadsoldier** [comment]
Rebased...

**Comment by hay-kot** [comment]
Thanks for those clarifications! Really helped me understand how this fits. I absolutely think we can move forward with this and get it cleaned up for release. The things I see that still need to get done are...

1. Documentation on implementation
2. Registration/Creation flow for SSO Auths
3. Integration and/or Unit Testing for all the new user auth flows...

Anything else?

**Comment by hay-kot** [comment]
> Can you point me to the place this should go? Should probably be an extension to the docs for the current authentication flow?

I would probably just add a new page for it 

https://github.com/hay-kot/homebox/tree/main/docs/docs

> Do you mean the automatic creation of Homebox users as mentioned in my previous comment?

Yes

> Could you point me to tests for the current flow please, so I would try to build something similar for the new flow?

I believe most of that is done via units tests and the user flow tests are here

https://github.com/hay-kot/homebox/blob/main/frontend/lib/api/__test__/user/group.test.ts

**Comment by verybadsoldier** [comment]
Sorry for the delay, I will be back on this shortly.

**Comment by icsy7867** [comment]
Just wanted to chime in here while I had a second and point out a couple of other services that can use header authentication as well:
https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/auth-proxy/
https://demo.netbox.dev/static/docs/administration/authentication/

From my depoyments using ADFS/SAML using mellon as a reverse proxy and using header auth, I think this would be a lot more versatile if there was a variable we could set (especially in the docker container using an environment variable) where we could manually set the header name used in authentication. 

I.E Something like this:
```
docker run -d \
  --name homebox \
  --restart unless-stopped \
  --publish 3100:7745 \
  --env TZ=Europe/Bucharest \
  --env auth_proxy=true \
  --env header_name="X-WEBAUTH-USER" \
  --volume /path/to/data/folder/:/data \
  ghcr.io/hay-kot/homebox:latest
```

But I am following this and excited about the addition!

**Comment by verybadsoldier** [comment]
So, I added the automatic creation of a user if not yet existing. As the password for the new user, a randomly generated string is used. Not meant to be used for actual logins.

I am a bit unsure how to handle group assignments when creating users. It seems groups are only used for assigning currencies right now. Also, I could not find a way in the UI to assign users to groups or to manage groups.
So, Authelia sends HTTP header `Remote-Groups`. But I expect those group names to be often application-specific, so I don't know how useful this is to use. I implemented it now the same way as users are created regularly by passing the empty string as group. So `Remote-Groups` HTTP header is not used here. Do you agree to this?

Making the HTTP header names configurable as @icsy7867 suggested is a good idea, I think. I will implement that.


PS.
Sorry for the delay!

**Comment by verybadsoldier** [comment]
Fixed CI tests and added documentation.

Still struggling with the tests. It's a bit difficult to test because an SSO service would be needed between client and backend.
My only idea would be to extend the API by adding a parameter `headers` to allow for the test client to send custom HTTP headers to emulate the headers generated by the SSO service. Could be done like this by adding a parameter to the login function:
```
export class PublicApi extends BaseAPI {
...

  public login_sso_header(headers: Map<string, string>) {
    return this.http.post<string, TokenResponse>({
      url: route("/users/login-sso-header"),
      headers: headers,
    });
  }
```

**Comment by verybadsoldier** [comment]
Okay, I did the unit tests as described by adding the ability to inject custom HTTP headers for the sake of testing.

From my point, I think all aspects are covered now (functions, docs, tests). So, feel free to notify me when changes are needed.

File: Taskfile.yml [hunk 1]
```diff
@@ -127,7 +127,7 @@ tasks:
     desc: Runs end-to-end test on a live server (only for use in CI)
     cmds:
       - cd backend && go build ./app/api
-      - backend/api &
+      - HBOX_OPTIONS_HEADER_SSO_ENABLED=1 HBOX_OPTIONS_HEADER_SSO_ALLOWED_IP=127.0.0.1 backend/api &
```
Comment (hay-kot) [inline Taskfile.yml:130]: Can you change this to use the task env config?

See here https://taskfile.dev/api/#task

File: backend/internal/core/services/service_user.go [hunk 2]
```diff
@@ -193,6 +193,17 @@ func (svc *UserService) Login(ctx context.Context, username, password string, ex
 	return svc.createSessionToken(ctx, usr.ID, extendedSession)
 }
 
+func (svc *UserService) LoginWithoutPassword(ctx context.Context, username string, extendedSession bool) (UserAuthTokenDetail, error) {
+	usr, err := svc.repos.Users.GetOneEmail(ctx, username)
+	if err != nil {
+		// SECURITY: Perform hash to ensure response times are the same
+		hasher.CheckPasswordHash("not-a-real-password", "not-a-real-password")
+		return UserAuthTokenDetail{}, ErrorInvalidLogin
+	}
```
Comment (hay-kot) [inline backend/internal/core/services/service_user.go:202]: Probably don't need to do the check password hash here since we're not doing that anyways, so we don't need to prevent any kind of timing attack

File: backend/internal/sys/config/conf.go [hunk 3]
```diff
@@ -26,8 +26,13 @@ type Config struct {
...
 type Options struct {
-	AllowRegistration    bool `yaml:"disable_registration" conf:"default:true"`
-	AutoIncrementAssetID bool `yaml:"auto_increment_asset_id" conf:"default:true"`
+	AllowRegistration       bool `yaml:"disable_registration" conf:"default:true"`
+	AutoIncrementAssetID    bool `yaml:"auto_increment_asset_id" conf:"default:true"`
+	HeaderSSOEnabled        bool `yaml:"header_sso_enabled" conf:"default:false"`
+	HeaderSSOAllowedIP      string `yaml:"header_sso_allowed_ip" conf:"default:0.0.0.0"`
+	HeaderSSOAutoRegister   bool `yaml:"header_sso_autoregister" conf:"default:true"`
+	HeaderSSOHeaderName     string `yaml:"header_sso_header_name" conf:"default:Remote-Name"`
+	HeaderSSOHeaderEmail    string `yaml:"header_sso_header_email" conf:"default:Remote-Email"`
```
Comment (hay-kot) [inline backend/internal/sys/config/conf.go:35]: Do we need to do any kind of validation to make sure that these values are valid? If possible it would be nice to prevent to user from using "bad" values and we can just log the error and exit. 

not super necessary, but can be nice in avoiding issues with mis-configuration.

File: frontend/composables/use-auth-context.ts [hunk 4]
```diff
@@ -122,6 +127,18 @@ class AuthContext implements IAuthContext {
     return r;
   }
 
+  async login_sso_header(api: PublicApi) {
```
Comment (hay-kot) [inline frontend/composables/use-auth-context.ts:130]: should be `loginSSOHeader` for conform the frontend casing.

File: backend/app/api/handlers/v1/v1_ctrl_auth.go [hunk 5]
```diff
@@ -89,6 +90,80 @@ func (ctrl *V1Controller) HandleAuthLogin() errchain.HandlerFunc {
...
+			//
+			/* TODO: decide how to handle group information provided by HTTP header
+			// if groups are provided, they will be comma-separated. take only the first group
+			var groups = r.Header.Get("Remote-Groups")
+			var groupArr = strings.Split(groups, ",")
+			groupTok := ""
+			if len(groupArr) > 0 {
+				groupTok = groupArr[0]
+			}
+			*/
```
Comment (hay-kot) [inline backend/app/api/handlers/v1/v1_ctrl_auth.go:135]: Where are you at with the group support? Currently every one will get their own group right now, correct?

**Review by hay-kot — CHANGES_REQUESTED** [review]
Overall looks _really_, _really_ good. Really appreciate the documentation as well. There's a couple small things that need to get cleaned up - and I need some clarification on how we're handling groups for SSO but this is super close.

File: backend/internal/sys/config/conf.go (hunk 3 above)
Comment (verybadsoldier) [reply backend/internal/sys/config/conf.go:35]: Hm, right now I can't think of any meaningful validation being done here. I thought about `HeaderSSOAllowedIP` but I am not sure if it always has to be an actual IP address. So I renamed that `HeaderSSOAllowedAddr`.
If you have ideas what could be validated here, I will be happy to add.

File: backend/app/api/handlers/v1/v1_ctrl_auth.go (hunk 5 above)
Comment (verybadsoldier) [reply backend/app/api/handlers/v1/v1_ctrl_auth.go:135]: Regarding groups, please see my previous comment about it:
https://github.com/hay-kot/homebox/pull/345#issuecomment-1522327160

TLDR:
Currently, it is handled the same way as regular registrations: every user is put into the group "Home".
I don't know if it makes sense to use the (possible) group information provided by the HTTP header because this would mean that different services (Homebox and others) need to agree on a common set of groups (group names). At least according to my understanding.

If you have a different idea how to handle the groups from the header, then I can implement that of course.

Comment (hay-kot) [reply backend/app/api/handlers/v1/v1_ctrl_auth.go:135]: > Currently, it is handled the same way as regular registrations: every user is put into the group "Home".

That isn't how it's currently handled though, unless I'm mis-understanding what you mean. Homebox is multi-tenant so each group has their own set of items, locations, etc... So if you're using SSO I'd assume you'd want some way for users to all be in the same group. I could see this going two ways,

1) Use a header to specify the group and either Add the user if the group exists, or create the group if it doesn't
2) Default all SSO users to some predefined group name (possibly set via env)

Again, maybe I'm just missing obvious...

Comment (verybadsoldier) [reply backend/app/api/handlers/v1/v1_ctrl_auth.go:135]: First, pretty sure I am the confused one here ;)

Ok, I didn't know about this multi-tenant thing, good to know, thanks. Anyway, without understanding what groups exactly do, I tried to handle groups in the SSO-Autoregistration the same way as (I thought) it works for regular user registration.
From what I saw in debugging, when doing a regular registration, then `HandleUserRegistration` is called with an empty group string (`""`) (since you cannot select/specify a group when registering?). Then when `RegisterUser` is called with the empty string as `GroupToken`, it falls back to a default group `Home`:
```
	switch data.GroupToken {
	case "":
		log.Debug().Msg("creating new group")
		creatingGroup = true
		group, err = svc.repos.Groups.GroupCreate(ctx, "Home")
		if err != nil {
			log.Err(err).Msg("Failed to create group")
			return repo.UserOut{}, err
		}
	default:
		log.Debug().Msg("joining existing group")
		token, err = svc.repos.Groups.InvitationGet(ctx, hasher.HashToken(data.GroupToken))
		if err != nil {
			log.Err(err).Msg("Failed to get invitation token")
			return repo.UserOut{}, err
		}
		group = token.Group
	}
```

So, that was as my assessment for it, so I also just used the emptry string as `GroupToken` for SSO-Autoregister. Just out of interest: did I miss something here?

Anyway, I like your second idea to make it configurable via environment variable and will go that way!
> Default all SSO users to some predefined group name (possibly set via env)

Comment (verybadsoldier) [reply backend/app/api/handlers/v1/v1_ctrl_auth.go:135]: Sorry, me again. This group thing still confuses me a bit, but after further studying the code, I think I had some insights. Also, I realized what I wrote above is not quite correct. The point is that a group is not identified by its name but by a UUID, right? So, there can be multiple groups with the name "Home" and for every normal user registration such a group is created for the new user (I remember you mentioned that above somewhere, and now I understand it).

> Default all SSO users to some predefined group name (possibly set via env)

Do you mean (similar to regular registration) creating a new group for every SSO auto registration and use this predefined group name (instead of "Home")?

Wouldn't it be sensible to handle it the same way as regular user registration with empty group token? So, just creating a new group named "Home" for every SSO user? That would be pretty consistent then with the regular registration. Basically how it is currently ;)

Comment (verybadsoldier) [reply backend/app/api/handlers/v1/v1_ctrl_auth.go:135]: So, I see three options to handle groups when auto-registering an SSO user:
1. Create a new group for every user (similar to regular registration) using a configurable default group name (I think that is what you proposed)
2. Handle it the same way as regular registration: create new group named "Home" for every user
3. Join the new user to an existing group by a configurable group UUID

I am fine with any of these. I am not entirely sure if I understood you correctly, so I would wait for confirmation before going on.

Comment (verybadsoldier) [reply backend/app/api/handlers/v1/v1_ctrl_auth.go:135]: Any opinions about this? Maybe we can finish this once this group thing is solved.

**Comment by verybadsoldier** [comment]
@hay-kot 
So, is this dead or are you planning to come back to this?

**Comment by agc93** [comment]
Is there any update on this? Seems like most of the work is done and getting support for SSO seems like a big win.

**Comment by markvdb** [comment]
I am exploring homebox for our non-profit. SSO functionality would be a big argument. Following along.

File: docs/docs/quick-start.md [hunk 6]
```diff
@@ -87,6 +92,11 @@ volumes:
         --debug-port/$HBOX_DEBUG_PORT                                            <string>  (default: 4000)
         --options-allow-registration/$HBOX_OPTIONS_ALLOW_REGISTRATION            <bool>    (default: true)
         --options-auto-increment-asset-id/$HBOX_OPTIONS_AUTO_INCREMENT_ASSET_ID  <bool>    (default: true)
+        --options-header-sso-enabled/$HBOX_OPTIONS_HEADER_SSO_ENABLED            <bool>    (default: false)
+        --options-header-sso-allowed_ip/$HBOX_OPTIONS_HEADER_SSO_ALLOWED_IP      <string>
+        --options-header-sso-autoregsiter/$HBOX_OPTIONS_HEADER_SSO_AUTOREGISTER  <bool>    (default: true)
```
Comment (baod-rate) [inline docs/docs/quick-start.md:97]: typo:

```suggestion
        --options-header-sso-autoregister/$HBOX_OPTIONS_HEADER_SSO_AUTOREGISTER  <bool>    (default: true)
```

File: docs/docs/quick-start.md [hunk 7]
```diff
@@ -87,6 +92,11 @@ volumes:
...
+        --options-header-sso-allowed_ip/$HBOX_OPTIONS_HEADER_SSO_HEADER_EMAIL    <string>  (default: Remote-Email)
```
Comment (baod-rate) [inline docs/docs/quick-start.md:98]: another typo:

```suggestion
        --options-header-sso-header-email/$HBOX_OPTIONS_HEADER_SSO_HEADER_EMAIL  <string>  (default: Remote-Email)
```

File: docs/docs/quick-start.md [hunk 8]
```diff
@@ -87,6 +92,11 @@ volumes:
...
+        --options-header-sso-allowed_ip/$HBOX_OPTIONS_HEADER_SSO_HEADER_NAME     <string>  (default: Remote-Name)
```
Comment (baod-rate) [inline docs/docs/quick-start.md:99]: another typo:

```suggestion
        --options-header-sso-header-name/$HBOX_OPTIONS_HEADER_SSO_HEADER_NAME    <string>  (default: Remote-Name)
```

**Comment by hay-kot** [comment]
Auth is kind of broken at the moment. I need to do some significant rewriting of it to fix some broken user paths. Once I get that figured out I'll circle back on this and look at finishing the implementation.

**Comment by hay-kot** [comment]
Provider interface for extending auth merged in

- https://github.com/hay-kot/homebox/pull/663

**Comment by quentinmit** [comment]
Why does this (and the `provider` interface) only apply to the `login` route? The headers should be checked on every request, and should take the place of a separate auth token (because otherwise your login state could get out of sync between HomeBox and your SSO provider).

**Comment by hay-kot** [comment]
Closing this for now. I don't have time right now to walk through all the changes required here. It's not something I'm going to be working on as I see this more as an business feature. If anyone wants to sponsor this work, I'd be happy to talk about that. Thanks!

**Comment by LIRIKKER** [comment]
Interesting you see this as a business feature when may people who are running a home lab using docker would be running some sort of SSO whether that’s through Authentik, Authelia, Keycloak etc. For me personally (and I, sure others agree), it’s the only thing stopped me using this over snipe-it.

could look in the interim supporting passkeys?
